libs/shared/utils/src/lib/delay.ts

```diff
@@ -7,9 +7,36 @@
  * Creates a promise that resolves after a specified delay.
  *
  * @param {number} [ms=500] The delay in milliseconds.
+ * @param  {AbortSignal} [signal] Optional AbortSignal to cancel the delay
  * @returns {Promise<void>} A promise that resolves after the delay.
  */
 
-export const delay = (ms: number = 500): Promise<void> => {
-  return new Promise((resolve) => setTimeout(resolve, ms))
-}
+export const delay = (
+  ms: number = 500,
+  signal?: AbortSignal
+): Promise<void> => {
+  return new Promise<void>((resolve, reject) => {
+    if (signal?.aborted) {
+      resolve();
+      return
+    }
+
+    const timer = setTimeout(() => {
+      cleanup();
+      resolve();
+    }, ms);
+
+    const onAbort = () => {
+      console.log('Delay aborted');
+      clearTimeout(timer);
+      cleanup();
+      resolve();
+    };
+
+    const cleanup = () => {
+      signal?.removeEventListener('abort', onAbort);
+    };
+
+    signal?.addEventListener('abort', onAbort);
+  });
```

Comment:
I'm not sure this is valid approach. The `delay` method is widely used, but the new logic is used only in one place. Instead of this, I'd revert the previous implementation and just use it like this:

```TSX
delay(signal.aborted ? 0 : 500)
```

Looking at its implementation...

```TSX
      if (signal.aborted) {
        return new AppError(ApiFileTransferError.Aborted, "Aborted")
      }

      if (!ok) {
        return error
      }

      await delay(500, signal)
```

... you can clearly see, that if signal is aborted, the delay function probably won't be called at all as there's a `return` before, and even if it would be called somehow, the solution proposed above should resolve the issue.

Reply:
Discussed and understood 🙂